### DISCUSSION TITLE
Great FOSS app and great dev work!

### DISCUSSION THREAD

**Discussion by farfromrefug** [CatimaLoyalty/Android#1909] [General]
Hi,
First i wanted to say thank you for that great app.
I am actually developing my own card wallet app [here](https://github.com/Akylas/OSS-DocumentScanner). The main difference for now is the fact that that app (not yet released) is based on the same code as OSS DocumentScanner and thus brings document detection, OCR and other feature. It will also be an iOS app.

I wanted to write that message to you to thank you because i took and translated some code from your app (and i might again for some of your great ideas!). For now i used your code to add android shortcuts and the design for your from [BarcodeSelectorActivity.java](https://github.com/CatimaLoyalty/Android/blob/main/app/src/main/java/protect/card_locker/BarcodeSelectorActivity.java) (still thinking about another design more minimalist).

I hope it is ok with you. And if you want i can report here from any other part i take inspiration from.

Thanks

**Comment by TheLastProject** [comment]
Hey @farfromrefug,

Thanks for reaching out, sounds like a cool project 🙂 People often ask me for an iOS version of Catima so even though it won't be exactly the same I'm sure this will be useful to many people. It would be great if it could support catima.app links for easy sharing at some point so it could really be an "ecosystem" 🙂

About your question: While I don't personally mind as long the license is respected (which also means proper attribution for the code you chose to use/translate), I do feel I should warn you about the fact that Catima is a GPL project. Of course, IANAL TINLA in this message, but here is how I understand it:

Given Catima is under the GPLv3+ and that being a strong copyleft license means that including any GPL code makes the combination of your MIT code and the Catima code a "combined work" which the GPL says should be GPL licensed as a whole. Sadly, due to Apple restricting users what they can do with their own devices the Apple App Store's Terms of Service is incompatible with the GPL, meaning you are technically not allowed to release a project that contains GPL code on the Apple App Store.

While I am personally not interested in enforcing my copyright against clearly well-intentioned Open Source projects like yours I feel I should warn you that because I don't require anyone to sign a CLA every contributor keeps full copyright over their code. So if anyone else decides to complain to Apple, there is nothing I could do about it.

This is probably way more negative a message than you were hoping to hear, but I'd rather you know about the risks up-front: the GPL and the Apple App Store are not compatible so using Catima code for an iOS project is legally problematic and risky, although I have no plans to ever complain to Apple (I don't think it's in the best interest of anyone).

I think @obfusk can explain the risks in more detail than me if things are unclear.

If you see ways to improve Catima or need help integrating catima.app URLs or whatever so the apps can work together better please feel free to reach out, sharing knowledge is always great 🙂

**Reply by farfromrefug** [reply]
@TheLastProject Thanks for the reply. Did not think about the licensing stuff. And indeed seems like Apple is not compatible with GPL code. I ll have to keep that in mind. Should not be a problem with what i have done so far:
* BarcodeSelectorActivity: i just took inspiration on the design. The whole code is in another language so no issue there.
* the notification part: small part and i am going to rewrite it anyway. I will keep your logic though as i like it!

But i ll keep all that in mind for the future and will only use your code as inspiration. I dont want to go against any license!
I will definitely add support for catima.app URLs and try to make our apps work together as much as possible.

Thanks a lot for your answer i really appreciate it

**Comment by Pozi09** [comment]
You guys are cool, it was interesting to read

**Comment by kintaro1981** [comment]
any news about your app @farfromrefug ?

**Reply by farfromrefug** [reply]
@kintaro1981 the app is out I have put the link in my first post. 
Right now I am.on a 6months break but Dev will start again as I come home.

**Reply by axgd-code** [reply]
@TheLastProject @farfromrefug A great feature idea for Catima and OSS CardWallet: the ability to export/import or read all cards via a file.
Just like I do with my .kdbx file, I dream of a single file containing all my cards — compatible with the best open-source apps 👏
I would be very easy to sync my Android and iOS devices

**Reply by TheLastProject** [reply]
Catima already can export all cards to a single zip file. JSON doesn't seem very viable to store everything in as Catima cards can contain images.

**Reply by axgd-code** [reply]
My main goal was to support projects like OSSCardWallet with a lightweight format that allows quick integration. I assumed images wouldn't be included, which doesn't seem like a major issue in this context.  Otherwise, it would be possible to store them as base64 if needed.

Also, syncing through a single file would probably be simpler than using a zip archive.

I was thinking of a format like this:
`
[
  {
    "card": "walmart",
    "description": "My Walmart card",
    "format": "qrcode",
    "code": "xytzmzlzl"
  }
]
`

**Reply by TheLastProject** [reply]
We have had this export format for several years and some apps already import from it: https://catimaloyalty.readthedocs.io/en/latest/docs/export_format/

I'd be hesitant to break existing importers in other apps and breaking importing Catima backups in older releases unless there isn't really a better way :)
